src/internal/storage/chunk/reader.go

```diff
@@ -11,24 +11,44 @@ import (
 
 // Reader reads data from chunk storage.
 type Reader struct {
-	ctx      context.Context
-	client   Client
-	memCache kv.GetPut
-	dataRefs []*DataRef
+	ctx         context.Context
+	client      Client
+	memCache    kv.GetPut
+	dataRefs    []*DataRef
+	offsetBytes int64
 }
 
-func newReader(ctx context.Context, client Client, memCache kv.GetPut, dataRefs []*DataRef) *Reader {
-	return &Reader{
+type ReaderOption func(*Reader)
+
+func WithOffsetBytes(offsetBytes int64) ReaderOption {
+	return func(r *Reader) {
+		r.offsetBytes = offsetBytes
+	}
+}
+
+func newReader(ctx context.Context, client Client, memCache kv.GetPut, dataRefs []*DataRef, opts ...ReaderOption) *Reader {
+	r := &Reader{
 		ctx:      ctx,
 		client:   client,
 		memCache: memCache,
 		dataRefs: dataRefs,
 	}
+	for _, opt := range opts {
+		opt(r)
+	}
+	return r
 }
 
 // Iterate iterates over the data readers for the data references.
 func (r *Reader) Iterate(cb func(*DataReader) error) error {
 	for _, dataRef := range r.dataRefs {
+		if dataRef.SizeBytes < r.offsetBytes {
+			r.offsetBytes -= dataRef.SizeBytes
+			continue
+		}
+		dataRef.OffsetBytes += r.offsetBytes
+		dataRef.SizeBytes -= r.offsetBytes
+		r.offsetBytes = 0
```

Comment:
I think there might be an issue with this.  We shouldn't be modifying the datarefs, or writing to the Reader variables.  We need to be able to call iterate more than once.  The Reader is supposed to be like a specification for the iteration.

I think we need to change `chunk.DataReader` to take an offset, and use that offset in `Get`.  Then in this loop we skip the refs that are before, use the DataReader with the offset for the overlap, and then just iterate as usual for the rest.

Reply:
good catch! that makes sense 👍